benchmark/benchmark_revisions.py

```diff
@@ -0,0 +1,119 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Benchmarking tool for different commits
+"""
+# pylint: disable=subprocess-run-check
+import argparse
+import os
+import subprocess
+
+# ANSI escape sequences for terminal colors
+Colors = {
+    "red": "\033[31m",
+    "yellow": "\033[33m",
+    "blue": "\033[34m",
+    "magenta": "\033[95m",
+}
+RESET = "\033[0m"
+
+
+def col(text, color):
+    """Wraps the given text in color ANSI sequences.
+
+    Args:
+        text  (str): text to print
+        color (str): ANSI color code
+
+    Returns:
+        str: text wrapped with ANSI codes
+    """
+    return Colors[color] + text + RESET
+
+
+class cd:
+    """Context manager for changing the current working directory"""
+
+    def __init__(self, newPath):
+        self.newPath = os.path.expanduser(newPath)
+        self.savedPath = None
+
+    def __enter__(self):
+        self.savedPath = os.getcwd()
+        os.chdir(self.newPath)
+
+    def __exit__(self, etype, value, traceback):
+        os.chdir(self.savedPath)
+
+
+# benchmarking tool version
+__version__ = "0.1.0"
+
+
+def cli():
+    """Parse the command line arguments, perform the requested action.
+    """
+    parser = argparse.ArgumentParser(description="PennyLane benchmarking tool for revisions")
+    parser.add_argument(
+        "-r",
+        "--revisions",
+        type=lambda x: x.split(","),
+        help="comma-separated list of revisions to run the benchmark on",
+    )
+
+    args, unknown_args = parser.parse_known_args()
+
+    revisions_directory = os.path.join(os.path.dirname(os.path.realpath(__file__)), "revisions")
+    if not os.path.exists(revisions_directory):
+        os.mkdir(revisions_directory)
+
+    for revision in args.revisions:
+        print(">>> Running benchmark for revision {}".format(col(revision, "red")))
+        pl_directory = os.path.join(revisions_directory, revision)
+
+        # We first make sure we get the latest version of the desired revision
+        if os.path.exists(pl_directory):
+            print(">>> Revision found locally, updating...")
+            with cd(pl_directory):
+                # Check if we're on a detached HEAD (i.e. for version revisions)
+                res = subprocess.run(
+                    "git rev-parse --abbrev-ref --symbolic-full-name HEAD",
+                    capture_output=True,
+                    check=True,
+                )
+
+                if "HEAD" not in str(res.stdout):
+                    subprocess.run("git checkout {} -q".format(revision))
+                    subprocess.run("git pull -q")
+        else:
+            print(">>> Revision not found locally, cloning...")
+            os.mkdir(pl_directory)
+            with cd(pl_directory):
+                subprocess.run("git clone https://www.github.com/xanaduai/pennylane . -q")
+                subprocess.run("git checkout {} -q".format(revision))
```

Comment:
Some sort of minimal error handling would be nice to have here. If `revision` is not valid, an `error: pathspec <revision> did not match any file(s) known to git` message is printed (even with `-q`).

Reply:
Thanks, I didn't think of this! Right now I try to check out the revision and if this fails I delete the temporary data. I need to download the whole source to do this though... There are git plumbing tools for checking if a tag or branch exists on a remote repo (`git ls-remote`) but that doesn't work for commits. 

I think I can speed this up a bit though by just copying a git repo and now cloning it each time

Reply:
I implemented the copying and it helps the performance a lot! It's not perfect, but for our purposes it should do the trick.